### DISCUSSION TITLE
Desktop support

### DISCUSSION THREAD

**Discussion by iampopal** [davidmartos96/sqflite_sqlcipher#45] [General]
I am using https://pub.dev/packages/sqflite_common_ffi currently as windows database, can we have suport of this package in windows operating system.

**Comment by davidmartos96** [comment]
@iampopal You can use sqlcipher with sqflite_common_ffi. Check out this: https://github.com/davidmartos96/sqflite_sqlcipher/issues/28#issuecomment-733636725

I don't have plans to support it via this package because this is just a wrapper of the native iOS and Android libraries.

**Comment by alextekartik** [comment]
Thanks @davidmartos96 nice trick I was not aware of. Your comment could definitely make it into the doc. I would link to it if you have it somewhere or copy/paste it in sqflte_common_ffi if you don't mind.

**Comment by davidmartos96** [comment]
@alextekartik The only place is that issue thread. You can copy the snippet, I don't mind.

**Comment by iampopal** [comment]
@davidmartos96 Thank you for the comment, but I am not able to load sqlcipher.dll file in windows 

Asynchronous error
Invalid argument(s): Failed to load dynamic library 'pathTo\sqlcipher.dll': 193

**Comment by iampopal** [comment]
I have download sqlcipher.dll file from 
https://www.dll4free.com/sqlcipher.dll.html
website for windows

**Comment by iampopal** [comment]
This is the code for loading sqlcipher.dll

```dart
  static DynamicLibrary sqlcipherOpen() {
    if (Platform.isLinux || Platform.isAndroid) {
      return DynamicLibrary.open('libsqlcipher.so');
    }
    if (Platform.isMacOS || Platform.isIOS) {
      return DynamicLibrary.open('/usr/lib/libsqlcipher.dylib');
    }
    if (Platform.isWindows) {
      return DynamicLibrary.open(join(
          'pathTo',
          'sqlcipher.dll'));
    }

    throw UnsupportedError(
        'moor_ffi does not support ${Platform.operatingSystem} yet');
  }
```

**Comment by davidmartos96** [comment]
@iampopal I wouldn't download any dll from those kind of websites, watch out for that. I'd recommend that you follow the compilation instructions and do it yourself.
You can check the following thread, in particular this comment. It could help you out regarding compilation on Windows
https://github.com/davidmartos96/sqflite_sqlcipher/issues/20#issuecomment-697051979

**Comment by davidmartos96** [comment]
@iampopal Also double check the path of the dll when loading it from Dart
The following `join('pathTo', 'sqlcipher.dll')` generates `pathTo/sqlcipher.dll` which I don't think it will be a path on your machine.

**Comment by iampopal** [comment]
> @iampopal I wouldn't download any dll from those kind of websites, watch out for that. I'd recommend that you follow the compilation instructions and do it yourself.
> You can check the following thread, in particular this comment. It could help you out regarding compilation on Windows
> [#20 (comment)](https://github.com/davidmartos96/sqflite_sqlcipher/issues/20#issuecomment-697051979)

Thank you for your quick replay, i will watch it.

**Comment by iampopal** [comment]
> @iampopal Also double check the path of the dll when loading it from Dart
> The following `join('pathTo', 'sqlcipher.dll')` generates `pathTo/sqlcipher.dll` which I don't think it will be a path on your machine.

pathTo is the full path the file folder and it's totally correct. 
the issue is I am not able to find the correct version of sqlciper.dll

**Comment by davidmartos96** [comment]
@iampopal    You can try using an absolute path for loading it first. I know from experience that loading with a relative path may be different depending on how you run the app. So maybe it's simply that the relative path is incorrect. After you know the DLL is correct you can investigate the relative path.

**Comment by iampopal** [comment]
@davidmartos96 I am getting this error 

Asynchronous error
Invalid argument(s): Failed to load dynamic library 'absolutePathTo\sqlcipher.dll': 193

when using absolute path

**Comment by iampopal** [comment]
absolutePathTo 
Is long path of my Windows pc to the file

**Comment by davidmartos96** [comment]
@iampopal If you are using the downloaded dll I wouldn't have my hopes up. Try with the compilation steps from https://github.com/daybson/sqlite_cipher_windows
This is the Community version of SQLCipher, so one needs the compile the libraries himself. If you need Commercial support or precompiled libraries you should look into the official website https://www.zetetic.net/sqlcipher/

**Comment by iampopal** [comment]
I have reviewed this folder and looked for sqlciper.dll but didn't found the sqlciper.dll file which used my many examples to load sqlcipher.dll

https://github.com/daybson/sqlite_cipher_windows/tree/master/sqlite%20cipher%20compiled/sqlcipher

**Comment by davidmartos96** [comment]
@iampopal It's probably the one called `sqlite.dll`. But if you need the latest version you would need to compile it yourself. I don't know the version that was used by the author of the repo back then.

**Comment by iampopal** [comment]
I did download the sqlite.dll file and it work with without password database

**Comment by iampopal** [comment]
@davidmartos96  
do u know is it possible to set the password for a database as it creates by the Moor package?

**Reply by davidmartos96** [reply]
Yes, you can do that. Check this out https://moor.simonbinder.eu/docs/other-engines/vm/ 
You should change the paths accordingly

**Reply by davidmartos96** [reply]
Sorry, wrong link. This one: https://moor.simonbinder.eu/docs/other-engines/encryption/#encrypted-version-of-moorffi

**Reply by davidmartos96** [reply]
For Desktop platforms you should update the following similar to what you had above.
```dart
import 'package:sqlite3/open.dart';

// call this method before using moor
void setupSqlcipher() {
  open.overrideFor(
      OperatingSystem.android, () => DynamicLibrary.open('libsqlcipher.so'));
}
```

**Reply by iampopal** [reply]
Thanks for this, i have able to open sqlite3.dll, now the issue is that when i pass password as 

```dart
LazyDatabase openConnection() {
  final dbFolder = 'C:\\Users\\Popal\\Documents\\exact\\';
  final dbPath = path.join(dbFolder, 'exact5.db');
  return LazyDatabase(() async {
    final file = File(dbPath);
    return VmDatabase(
      file,
      logStatements: true,
      setup: (db) {
        db.execute("PRAGMA KEY = 'Popal.dev'");
      },
    );
  });
}
````
i get the follwing error: 

SqliteException(26): file is not a database, file is not a database (code 26)

**Reply by davidmartos96** [reply]
If the database already exists it may be that you are using the wrong password. To verify that you are indeed using the sqlcipher library try running the following before the PRAGMA KEY

`print(await db.customSelect("PRAGMA cipher_version").get());`

**Comment by iampopal** [comment]
@davidmartos96 Thank you so much, I am able to run an encrypted database  using sqlite3.dll file from 
https://github.com/daybson/sqlite_cipher_windows/tree/master/sqlite%20cipher%20compiled/sqlcipher

**Comment by MobiliteDev** [comment]
I've made a package with the last community SQLCipher version (4.4.3) for Windows if you want.
https://pub.dev/packages/sqlcipher_library_windows
